benchmark/infrastructure/BenchmarkToString.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel February of 2023, schlegea@informatik.uni-freiburg.de)
+
+#include "../benchmark/infrastructure/BenchmarkToString.h"
+
+#include <absl/strings/str_cat.h>
+#include <absl/strings/str_replace.h>
+
+#include <sstream>
+
+#include "BenchmarkMeasurementContainer.h"
+#include "BenchmarkMetadata.h"
+#include "util/Exception.h"
+#include "util/Forward.h"
+
+namespace ad_benchmark {
+
+/*
+ * @brief Return a string of the form
+ * "#################
+ * # categoryTitle #
+ * #################"
+ */
+static std::string createCategoryTitle(std::string_view categoryTitle) {
+  // The bar above and below the title.
+  const size_t barLength = categoryTitle.size() + 4;
+  const std::string bar(barLength, '#');
+
+  return absl::StrCat(bar, "\n# ", categoryTitle, " #\n", bar);
+}
+
+// ___________________________________________________________________________
+std::string addIndentation(const std::string_view str,
+                           const size_t& indentationLevel) {
+  // An indention level of 0 makes no sense. Must be an error.
+  AD_CONTRACT_CHECK(indentationLevel > 0);
+
+  // The indention symbols for this level of indention.
+  std::string indentationSymbols{""};
+  indentationSymbols.reserve(outputIndentation.size() * indentationLevel);
+  for (size_t i = 0; i < indentationLevel; i++) {
+    indentationSymbols.append(outputIndentation);
+  }
+
+  // Add an indentation to the beginning and replace a new line with a new line,
+  // directly followed by the indentation.
+  return absl::StrCat(
+      indentationSymbols,
+      absl::StrReplaceAll(str,
+                          {{"\n", absl::StrCat("\n", indentationSymbols)}}));
+}
+
+// ___________________________________________________________________________
+std::string getMetadataPrettyString(const BenchmarkMetadata& meta,
+                                    std::string_view prefix,
+                                    std::string_view suffix) {
+  const std::string& metadataString = meta.asJsonString(true);
+  if (metadataString != "null") {
+    return absl::StrCat(prefix, metadataString, suffix);
+  } else {
+    return "";
+  }
+}
+
+/*
+@brief Return a string containing a titel version of `categoryName` and a string
+list representation of all the given category entries.
+Note: This function converts `CategoryType` objects by trying to cast them as
+`std::string`.
+*/
+template <typename CategoryType>
+static std::string categoryToString(
+    std::string_view categoryName,
+    const std::vector<CategoryType>& categoryEntries) {
+  return absl::StrCat(createCategoryTitle(categoryName), "\n",
+                      ad_utility::listToString(
+                          categoryEntries,
+                          [](const CategoryType& entry) {
+                            return static_cast<std::string>(entry);
```

Comment:
If you give all your types not only the `explicit operator std::string`, but also the `operator<<` for `ostream`, 
then you can get rid of all the `static_cast<...>` lambdas.

Reply:
Good idea, done in [ddcfb43](https://github.com/ad-freiburg/qlever/pull/1010/commits/ddcfb43f9fdde4fb178f7c212d941038bcfb1b1a).

However, now I got the weirdest assertion fault from `nlohmann::json`, when trying to run `BenchmarkExamples -p`. Considering, that I've never touched anything working with the json system in this specific issue, I'm kinda at a loss, where the hell it came from.
Stinks like undefined behavior to me.

Reply:
According to our workflow compilation with `gcc-13` is the problem not actually found with the json, but with my internal `operator<<` implementation:

```c++
static std::ostream& streamInserterOperatorImplementation(std::ostream& os, const auto& obj) {
  os << static_cast<std::string>(obj);
  return os;
}
```

Apparently, `static_cast<std::string>(obj)` is a dangeling pointer? Somehow?

Reply:
I need to see the exact error message,
And since the bug seems to be in the concrete `operator std::string` of some concrete type, I need to see that too.

Reply:
Okay, 
So clang crashes in the address sanitizer, in GCC11 the program crashes without much of an error (we don't have an address sanitizer activated there), and GCC13 has that warning (that is only semi-helpul in the release build, but stills points out that we have to further investigate).
Can you reproduce the crashes locally/also in Debug mode? There seems to be something wrong inside the `operator std::string` for `ResultTable`, but what exactly, has to be further determined.

Reply:
I can reproduce the `gcc11` crash, but that points to `nlohmann::json`, so it's not really helpfull.

Reply:
Well, that was an intense headache. Me and @joka921, mainly @joka921, managed to fix it however. See [d3cf4b7](https://github.com/ad-freiburg/qlever/pull/1010/commits/d3cf4b73a10011b1fb521f2ae15ac49f361830e3).

Turns out, calling `std::identify` on an `r-value` returns a r-value reference. Which, of course, points to an r-value, who already died, once the r-value reference was returned.